plutus-core/plutus-core/src/Bitwise.hs

```diff
@@ -0,0 +1,490 @@
+            let restLen :: CSize = fromIntegral len - mainLen
+            void $ case signum bigMove of
+              1 -> memcpy (plusPtr dst . fromIntegral $ restLen) src mainLen >>
+                   memcpy dst (plusPtr src . fromIntegral $ mainLen) restLen
+              _ -> memcpy (plusPtr dst . fromIntegral $ mainLen) src restLen >>
+                   memcpy dst (plusPtr src . fromIntegral $ restLen) mainLen
+          -- If we don't rotate by a multiple of 8, we have to construct new
+          -- bytes, rather than just copying over old ones. We do this in two
+          -- steps:
+          --
+          -- 1. Compute the 'read offset' into the source ByteString based on
+          -- the rotation magnitude and direction.
+          -- 2. Use that read offset to perform an (unchecked) bit lookup for an
+          -- entire 8-bit block, then construct the byte that results.
+          --
+          -- We can do the bytes in the result in any order using this method:
+          -- we choose to do it in traversal order.
+          _ -> for_ [0 .. len - 1] $ \j -> do
+                let start = (len - 1 - j) * 8
+                let dstByte = foldl' (addBit start displacement) zeroBits [0 .. 7]
+                poke (plusPtr dst j) dstByte
+      pure dst
+    bitLen :: Integer
+    bitLen = fromIntegral $ BS.length bs * 8
+    addBit :: Int -> Integer -> Word8 -> Integer -> Word8
+    addBit start displacement acc offset =
+      let oldIx = (offset + fromIntegral start + bitLen - displacement) `rem` bitLen in
+        if dangerousRead bs oldIx
+        then acc .|. (bit . fromIntegral $ offset)
+        else acc
+
+{-# NOINLINE shiftByteString #-}
+shiftByteString :: ByteString -> Integer -> ByteString
+shiftByteString bs i
+  -- Shifting by the number of bits, or more, would zero everything anyway,
+  -- regardless of direction. This also covers the empty ByteString case, as its
+  -- bit length is zero.
+  | abs i >= bitLen = BS.replicate (BS.length bs) zeroBits
+  -- Shifting an all-zero ByteString will not change it, regardless of
+  -- direction.
+  | isAllZero bs = bs
+  | otherwise = overPtrLen bs go
+  where
+    bitLen :: Integer
+    bitLen = fromIntegral $ BS.length bs * 8
+    go :: Ptr Word8 -> Int -> IO (Ptr Word8)
+    go src len = do
+      dst <- mallocBytes len
+      case len of
+        -- If we only have one byte, we an borrow from the Bits instance for
+        -- Word8.
+        1 -> do
+          srcByte <- peek src
+          let srcByte' = srcByte `shift` fromIntegral i
+          poke dst srcByte'
+        -- If we shift by a multiple of 8, we only need to move a contiguous
+        -- block of bytes, then clear what remains. This is much more efficient:
+        -- it would be nice if we had memset available, but at least the copy
+        -- can be done with memcpy.
+        _ -> case i `quotRem` 8 of
+          (bigMove, 0) -> do
+            let mainLen :: CSize = fromIntegral . abs $ bigMove
```

Comment:
We have quite a lot of `fromIntegral` calls that I'm not quite comfortable with. There's no explanation as for why they'd be sound.